### PR TITLE
Add support  for some new calls to the FossID API client

### PR DESCRIPTION
**Pull request by nnobelis**
See individual commits for details.

Theses calls will be used in the future to add snippets information to ORT data model.

See also https://github.com/oss-review-toolkit/ort/issues/3265.

### CODE REVIEW

File: clients/fossid-webapp/src/main/kotlin/Extensions.kt
Comment (sschuberth) [inline clients/fossid-webapp/src/main/kotlin/Extensions.kt:1]: Commit message nits:
- "Fossid-webapp" -> "fossid-webapp" (as it refers to the lower-case directory name)

File: clients/fossid-webapp/src/main/kotlin/Extensions.kt [hunk 1]
```diff
@@ -22,18 +22,23 @@
 package org.ossreviewtoolkit.clients.fossid
 
 import java.io.File
+import java.util.Base64
```
Comment (sschuberth) [inline clients/fossid-webapp/src/main/kotlin/Extensions.kt:25]: Note to myself: With Kotlin 1.8.20 we could use the [experimental Kotlin-native Base64 implementation](https://kotlinlang.org/docs/whatsnew-eap.html#experimental-support-for-base64-encoding-in-standard-library).

File: clients/fossid-webapp/src/main/kotlin/model/result/Snippet.kt [hunk 2]
```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (C) 2023 The ORT Project Authors (See <https://github.com/oss-review-toolkit/ort-server/blob/main/NOTICE>)
```
Comment (sschuberth) [inline clients/fossid-webapp/src/main/kotlin/model/result/Snippet.kt:2]: "See" -> "see"

Comment (sschuberth) [reply clients/fossid-webapp/src/main/kotlin/model/result/Snippet.kt:2]: And the URL needs to be https://github.com/oss-review-toolkit/ort/blob/main/NOTICE.

File: clients/fossid-webapp/src/main/kotlin/model/result/Snippet.kt [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.clients.fossid.model.result
+
+data class Snippet(
```
Comment (sschuberth) [inline clients/fossid-webapp/src/main/kotlin/model/result/Snippet.kt:22]: Are there some docs available for this class and its properties? Is really almost everything in this class optional, i.e. you cannot rely on the presence of properties?

File: clients/fossid-webapp/src/test/kotlin/FossIdClientReturnTypeTest.kt [hunk 4]
```diff
@@ -122,6 +124,23 @@ class FossIdClientReturnTypeTest : StringSpec({
         }
     }
 
+    "Snippets can be listed when there is some" {
```
Comment (sschuberth) [inline clients/fossid-webapp/src/test/kotlin/FossIdClientReturnTypeTest.kt:137]: Maybe shorter: "Existing snippets can be listed".

Comment (sschuberth) [reply clients/fossid-webapp/src/test/kotlin/FossIdClientReturnTypeTest.kt:137]: But I see other tests should then be reworded accordingly as well.

File: clients/fossid-webapp/src/main/kotlin/model/result/MatchedLines.kt [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
```
Comment (sschuberth) [inline clients/fossid-webapp/src/main/kotlin/model/result/MatchedLines.kt:20]: This needs to be moved above the `package`.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: clients/fossid-webapp/src/test/kotlin/FossIdClientReturnTypeTest.kt (hunk 4 above)
Comment (nnobelis) [reply clients/fossid-webapp/src/test/kotlin/FossIdClientReturnTypeTest.kt:137]: Can we leave them as if ? I don't intend to reword all tests now.

File: clients/fossid-webapp/src/main/kotlin/model/result/Snippet.kt (hunk 3 above)
Comment (nnobelis) [reply clients/fossid-webapp/src/main/kotlin/model/result/Snippet.kt:22]: No unfortunately as always with FossID, there is no documentation of the model available.

Maybe we can make the assumption `id`, `matchType`, `fieIdId` and `scanId` are not null, for the rest I don't know.

Please view https://github.com/oss-review-toolkit/ort/blob/34ca6c65da51415bfdf09c06223867fb140d2d32/clients/fossid-webapp/src/test/assets/return-type/mappings/apiphp-list_scan_snippets.json for example of the response.

Comment (sschuberth) [reply clients/fossid-webapp/src/main/kotlin/model/result/Snippet.kt:22]: > Maybe we can make the assumption `id`, `matchType`, `fieIdId` and `scanId` are not null

Yes, please make as many properties non-nullable as you confidently can, to simplify any upcoming code that makes use of these properties.

Comment (nnobelis) [reply clients/fossid-webapp/src/main/kotlin/model/result/Snippet.kt:22]: Ok I queried the snippets of several files of the Semver4j project and I make some changes to the class.

**Comment by sschuberth** [comment]
Can you please rebase this onto latest `main` once?

**Review by sschuberth — APPROVED** [review]

**Pull request merged**